File: src/main/java/org/luxons/sevenwonders/repositories/PlayerRepository.java
package org.luxons.sevenwonders.repositories;

import java.util.HashMap;
import java.util.Map;

import org.luxons.sevenwonders.errors.ApiMisuseException;
import org.luxons.sevenwonders.game.Player;
import org.springframework.stereotype.Repository;

@Repository
public class PlayerRepository {

    private Map<String, Player> players = new HashMap<>();

    public boolean contains(String username) {
        return players.containsKey(username);
    }

    public Player createOrUpdate(String username, String displayName) {
        if (players.containsKey(username)) {
            return update(username, displayName);
        } else {
            return create(username, displayName);
        }
    }

    private Player create(String username, String displayName) {
        Player player = new Player(username, displayName);
        players.put(username, player);
        return player;
    }

    private Player update(String username, String displayName) throws PlayerNotFoundException {
        Player player = find(username);
        player.setDisplayName(displayName);
        return player;
    }

    public Player find(String username) throws PlayerNotFoundException {
        Player player = players.get(username);
        if (player == null) {
            throw new PlayerNotFoundException(username);
        }
        return player;
    }

    public Player remove(String username) {
        Player player = players.remove(username);
        if (player == null) {
            throw new PlayerNotFoundException(username);
        }
        return player;
    }

    static class PlayerNotFoundException extends ApiMisuseException {
        PlayerNotFoundException(String username) {
            super("Player '" + username + "' doesn't exist");
        }
    }
}

File: backend/src/main/kotlin/org/luxons/sevenwonders/repositories/PlayerRepository.kt
package org.luxons.sevenwonders.repositories

import org.luxons.sevenwonders.errors.ApiMisuseException
import org.luxons.sevenwonders.lobby.Player
import org.springframework.stereotype.Repository
import java.util.HashMap

@Repository
class PlayerRepository {

    private val players = HashMap<String, Player>()

    operator fun contains(username: String): Boolean = players.containsKey(username)

    fun createOrUpdate(username: String, displayName: String): Player {
        return if (players.containsKey(username)) {
            update(username, displayName)
        } else {
            create(username, displayName)
        }
    }

    private fun create(username: String, displayName: String): Player {
        val player = Player(username, displayName)
        players[username] = player
        return player
    }

    private fun update(username: String, displayName: String): Player {
        val player = find(username)
        player.displayName = displayName
        return player
    }

    fun find(username: String): Player = players[username] ?: throw PlayerNotFoundException(username)

    fun remove(username: String): Player = players.remove(username) ?: throw PlayerNotFoundException(username)
}

internal class PlayerNotFoundException(username: String) :
    ApiMisuseException("Player '$username' doesn't exist")
